yjit/src/codegen.rs

```diff
@@ -3444,6 +3446,30 @@ fn jit_guard_known_klass(
         jit_mov_gc_ptr(jit, cb, REG1, sample_instance);
         cmp(cb, REG0, REG1);
         jit_chain_guard(JCC_JNE, jit, ctx, cb, ocb, max_chain_depth, side_exit);
+    } else if unsafe { known_klass == rb_cString && sample_instance.class_of() == rb_cString } {
+        assert!(!val_type.is_imm());
+
+        if val_type != Type::CString {
+
+            // Check that the receiver is a heap object
+            // TODO: can we remove this check? We know sample_instance is an rb_cString.
+            if !val_type.is_heap() {
+                add_comment(cb, "guard object is not immediate (for String)");
+                assert!(Qfalse.as_i32() < Qnil.as_i32());
+                test(cb, REG0, imm_opnd(RUBY_IMMEDIATE_MASK as i64));
+                jit_chain_guard(JCC_JNZ, jit, ctx, cb, ocb, max_chain_depth, side_exit);
+                cmp(cb, REG0, imm_opnd(Qnil.into()));
+                jit_chain_guard(JCC_JBE, jit, ctx, cb, ocb, max_chain_depth, side_exit);
+            }
+
+            add_comment(cb, &format!("guard object is un-subclassed String ({:?})", val_type));
+            let klass_opnd = mem_opnd(64, REG0, RUBY_OFFSET_RBASIC_KLASS);
+            mov(cb, REG1, uimm_opnd(unsafe { rb_cString }.into()));
+            cmp(cb, klass_opnd, REG1);
+            jit_chain_guard(JCC_JNE, jit, ctx, cb, ocb, max_chain_depth, side_exit);
+
+            ctx.upgrade_opnd_type(insn_opnd, Type::CString);
```

Comment:
Since this is the only line different from the general arm below modulo different comments, it might be better to have this in an if at the end of the general case and have this guard-elided if body be empty, something like the following:

```
...
else if val_type == Type::CString && known_klass == rb_cString {
    // guard elided because the context says we've already checked
    assert_eq!(sample_instance.class_of(), rb_cString, "context says class is exactly ::String");
} else {
    // general case
    // ...
    if known_klass == rb_cString {
        ctx.upgrade_opnd_type(insn_opnd, Type::CString);
    }
}
```

I think it's easier to see the elision path this way because one wouldn't need to skip past a big block of code in the current `if val_type != Type::CString` branch.

Reply:
Okay. I had wanted to preserve the "check type, check type, check type" top-level structure for the function. But I could put string inside the "else". You're right, it's nearly the same code.

Reply:
Ah... Yeah, this is definitely more elegant as a code change. Testing it out now.